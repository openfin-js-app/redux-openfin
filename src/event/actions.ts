import {EventConstant, appEvents, windowEvents} from './constants';

export type ActionDict = {
    [name:string]:EventConstant,
}

const appEventDictByName:ActionDict = {};
const appEventDictByType:ActionDict = {};
const windowEventDictByName:ActionDict = {};
const windowEventDictByType:ActionDict = {};

appEvents.forEach(oneEvent => {
    appEventDictByName[oneEvent.name] = oneEvent;
    appEventDictByType[oneEvent.type] = oneEvent;
});

windowEvents.forEach(oneEvent => {
    windowEventDictByName[oneEvent.name] = oneEvent;
    windowEventDictByType[oneEvent.type] = oneEvent;
});

export default {
    appEventDictByName,
    appEventDictByType,
    windowEventDictByName,
    windowEventDictByType,
};