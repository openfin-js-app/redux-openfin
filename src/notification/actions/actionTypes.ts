import makeType from '../../utils/makeType';

export const CREATE_NOTIFICATION = makeType('CREATE_NOTIFICATION');
export const CREATE_NOTIFICATION_RES = makeType('CREATE_NOTIFICATION_RES');