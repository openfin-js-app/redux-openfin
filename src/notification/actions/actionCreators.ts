import {Action} from 'redux-actions';
import createFSA, {ActionCreator} from '../../utils/createFSA';
import * as types from "../types";
import * as actionTypes from "./actionTypes";

export const createNotification:ActionCreator<Partial<types.CreateNotificationPayload>>
    =
createFSA<Partial<types.CreateNotificationPayload>>(
    actionTypes.CREATE_NOTIFICATION,
    (options:Partial<types.CreateNotificationPayload>) => <any>{options}
)