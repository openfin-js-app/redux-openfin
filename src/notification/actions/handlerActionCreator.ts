import {Action} from 'redux-actions';
import createFSA, {ActionCreator} from "../../utils/createFSA";
import * as types from "../types";
import * as actionTypes from "./actionTypes";

export const createNotificationRes:ActionCreator<types.CreateNotificationResPayload> = createFSA<types.CreateNotificationResPayload>(
    actionTypes.CREATE_NOTIFICATION_RES,
    (options:types.CreateNotificationResPayload) => <any>(options)
)


export const getCurrentRes:ActionCreator<types.NotificationGetCurrentResPayload>
    =
    createFSA<types.NotificationGetCurrentResPayload>(
        actionTypes.NOTIFICATION_GET_CURRENT_RES,
        (options:Partial<types.NotificationGetCurrentResPayload>) => <any>(options)
    )

export const closeRes:ActionCreator<types.NotificationCloseResPayload>
    =
    createFSA<types.NotificationCloseResPayload>(
        actionTypes.NOTIFICATION_CLOSE_RES,
        (options:Partial<types.NotificationCloseResPayload>) => <any>(options)
    )

export const sendMessageRes:ActionCreator<types.NotificationSendMsgResPayload>
    =
    createFSA<types.NotificationSendMsgResPayload>(
        actionTypes.NOTIFICATION_SEND_MSG_RES,
        (options:Partial<types.NotificationSendMsgResPayload>) => <any>(options)
    )

export const sendMessageToApplicationRes:ActionCreator<types.NotificationSendMsgToAppResPayload>
    =
    createFSA<types.NotificationSendMsgToAppResPayload>(
        actionTypes.NOTIFICATION_SEND_MSG_TO_APP_RES,
        (options:Partial<types.NotificationSendMsgToAppResPayload>) => <any>(options)
    )




// ----------------------notification events
// http://cdn.openfin.co/jsdocs/beta/fin.Notification.html#~onClick
// http://cdn.openfin.co/jsdocs/beta/fin.Notification.html#~onClose
// http://cdn.openfin.co/jsdocs/beta/fin.Notification.html#~onDismiss
// http://cdn.openfin.co/jsdocs/beta/global.html#errorCallback
// http://cdn.openfin.co/jsdocs/beta/fin.Notification.html#~onMessage
// http://cdn.openfin.co/jsdocs/beta/fin.Notification.html#~onShow

export const notificationOnClickRes:ActionCreator<types.NotificationOnClickResPayload> = createFSA<types.NotificationOnClickResPayload>(
    actionTypes.NOTIFICATION_ON_CLICK_RES,
    (options:types.NotificationOnClickResPayload) => <any>(options)
)
export const notificationOnCloseRes:ActionCreator<types.NotificationOnCloseResPayload> = createFSA<types.NotificationOnCloseResPayload>(
    actionTypes.NOTIFICATION_ON_CLOSE_RES,
    (options:types.NotificationOnCloseResPayload) => <any>(options)
)
export const notificationOnDismissRes:ActionCreator<types.NotificationOnDismissResPayload> = createFSA<types.NotificationOnDismissResPayload>(
    actionTypes.NOTIFICATION_ON_DISMISS_RES,
    (options:types.NotificationOnDismissResPayload) => <any>(options)
)
export const notificationOnErrorRes:ActionCreator<types.NotificationOnErrorResPayload> = createFSA<types.NotificationOnErrorResPayload>(
    actionTypes.NOTIFICATION_ON_ERROR_RES,
    (options:types.NotificationOnErrorResPayload) => <any>(options)
)
export const notificationOnMsgRes:ActionCreator<types.NotificationOnMsgResPayload> = createFSA<types.NotificationOnMsgResPayload>(
    actionTypes.NOTIFICATION_ON_MSG_RES,
    (options:types.NotificationOnMsgResPayload) => <any>(options)
)
export const notificationOnShowRes:ActionCreator<types.NotificationOnShowResPayload> = createFSA<types.NotificationOnShowResPayload>(
    actionTypes.NOTIFICATION_ON_SHOW_RES,
    (options:types.NotificationOnShowResPayload) => <any>(options)
)
