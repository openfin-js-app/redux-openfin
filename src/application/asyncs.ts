import {Action} from 'redux-actions';
import * as types from "./types";
import * as handlerActions from './actions/handlerActionCreator';

import createAsyncFun from '../utils/createAsyncFun'

import {
    NEW_APPLICAITON_ERROR_MSG,
    RESTART_ERROR_MSG,
    CLOSE_ERROR_MSG,
} from './types';

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#Application
export async function newApplication(action:Action<types.NewApplicationPayload>):Promise<Action<types.NewApplicationResPayload>>{
    const  options  = action.payload;
    return createAsyncFun<types.NewApplicationPayload,types.NewApplicationResPayload>(
        action,
        NEW_APPLICAITON_ERROR_MSG,
        handlerActions.newApplicationRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            let app = new fin.desktop.Application(options,
                (app:any)=>{
                    const responseAction = resActionCreator({app});
                    succCb(responseAction);
                },errCb);
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#.getCurrent

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#getWindow

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#.wrap

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#addEventListener

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#close
export async function close(action:Action<types.ClosePayload>):Promise<Action<types.CloseResPayload>>{
    const { force } = action.payload;
    return createAsyncFun<types.ClosePayload,types.CloseResPayload>(
        action,
        CLOSE_ERROR_MSG,
        handlerActions.closeRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            let application = fin.desktop.Application.getCurrent();
            application.close(force,
                ()=>{
                    const responseAction = resActionCreator({});
                    succCb(responseAction);
                },errCb);
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#getChildWindows

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#getInfo

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#getShortcuts

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#getTrayIconInfo

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#getZoomLevel

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#isRunning

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#removeEventListener

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#restart
export async function restart(action:Action<types.RestartPayload>):Promise<Action<types.RestartResPayload>>{
    return createAsyncFun<types.RestartPayload,types.NewApplicationResPayload>(
        action,
        RESTART_ERROR_MSG,
        handlerActions.restartRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            let application = fin.desktop.Application.getCurrent();
            application.restart(
                ()=>{
                    const responseAction = resActionCreator({});
                    succCb(responseAction);
                },errCb);
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#run

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#scheduleRestart

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#setShortcuts

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#setTrayIcon

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#setZoomLevel

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Application.html#terminate
