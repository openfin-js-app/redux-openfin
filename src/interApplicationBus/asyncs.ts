import {Action} from 'redux-actions';
import * as types from "./types";
import * as handlerActions from './actions/handlerActionCreator';

import wrapAsyncFun from '../utils/wrapAsyncFun'

export async function publish(action:Action<types.PublishPayload>):Promise<Action<types.PublishResPayload>>{
    const  { topic, message }  = action.payload;
    return wrapAsyncFun<types.PublishPayload,types.PublishResPayload>(
        action,
        handlerActions.publishRes,
        async (fin)=>{
            await fin.InterApplicationBus.publish(topic, message);
            return handlerActions.publishRes({});
        }
    );
}

export async function subscribe(action:Action<types.SubscribePayload>):Promise<Action<types.SubscribeResPayload>>{
    const  { source, topic, listener }  = action.payload;
    return wrapAsyncFun<types.SubscribePayload,types.SubscribeResPayload>(
        action,
        handlerActions.subscribeRes,
        async (fin)=>{
            await fin.InterApplicationBus.subscribe(source, topic, listener);
            return handlerActions.subscribeRes({});
        }
    );
}
