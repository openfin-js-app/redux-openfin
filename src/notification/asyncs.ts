import {Action} from 'redux-actions';
import * as types from "./types";
import * as handlerActions from './actions/handlerActionCreator';

import {initState} from '../init';

import createAsyncFun from '../utils/createAsyncFun'

import uuid from '../utils/uuid';

import {
    CREATE_NOTIFICATION_ERROR_MSG,
    NOTIFICATION_GET_CURRENT_ERROR_MSG,
    NOTIFICATION_CLOSE_ERROR_MSG,
    NOTIFICATION_SEND_MSG_ERROR_MSG,
    NOTIFICATION_SEND_MSG_TO_APP_ERROR_MSG,
} from './types';

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Notification.html#Notification
export async function createNotification(action:Action<types.CreateNotificationPayload>):Promise<Action<types.CreateNotificationResPayload>>{
    const  options  = action.payload;

    let userId = action.payload.userId;
    if (!userId){
        userId = uuid();
    }

    let finOptions:any = {...options};
    delete finOptions.userId;
    if (initState.store){
        const dispatch = initState.store.dispatch;
        finOptions.onClick = ()=>{
            dispatch(handlerActions.notificationOnClickRes({userId}));
        };
        finOptions.onClose = ()=>{
            dispatch(handlerActions.notificationOnCloseRes({userId}));
        };
        finOptions.onDismiss = ()=>{
            dispatch(handlerActions.notificationOnDismissRes({userId}));
        };
        finOptions.onError = (reason,errorObj)=>{
            dispatch(handlerActions.notificationOnErrorRes({userId,reason,errorObj}));
        };
        finOptions.onMessage = (message)=>{
            dispatch(handlerActions.notificationOnMsgRes({userId,message}));
        };
        finOptions.onShow = (successObj)=>{
            dispatch(handlerActions.notificationOnShowRes({userId,successObj}));
        };
    }

    return createAsyncFun<types.CreateNotificationPayload,types.CreateNotificationResPayload>(
        action,
        CREATE_NOTIFICATION_ERROR_MSG,
        handlerActions.createNotificationRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            let notification = new fin.desktop.Notification(finOptions,
                ()=>{
                    const responseAction = resActionCreator({
                        userId,
                        notification
                    });
                    succCb(responseAction);
                },errCb);
        }
    );
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Notification.html#.getCurrent
export async function getCurrent(action:Action<types.NotificationGetCurrentPayload>):Promise<Action<types.NotificationGetCurrentResPayload>>{
    return createAsyncFun<types.NotificationGetCurrentPayload,types.NotificationGetCurrentResPayload>(
        action,
        NOTIFICATION_GET_CURRENT_ERROR_MSG,
        handlerActions.getCurrentRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            const notification = fin.desktop.Notification.getCurrent();
            const responseAction = resActionCreator({notification})
            succCb(responseAction);
        }
    )
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Notification.html#close
export async function close(action:Action<types.NotificationClosePayload>):Promise<Action<types.NotificationCloseResPayload>>{
    return createAsyncFun<types.NotificationClosePayload,types.NotificationCloseResPayload>(
        action,
        NOTIFICATION_CLOSE_ERROR_MSG,
        handlerActions.closeRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            const notification = fin.desktop.Notification.getCurrent();
            const responseAction = resActionCreator({})
            notification.close(()=>{
                succCb(responseAction);
            });
        }
    )
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Notification.html#sendMessage
export async function sendMessage(action:Action<types.NotificationSendMsgPayload>):Promise<Action<types.NotificationSendMsgResPayload>>{

    const { payload: {
        message
    }} = action;

    return createAsyncFun<types.NotificationSendMsgPayload,types.NotificationSendMsgResPayload>(
        action,
        NOTIFICATION_SEND_MSG_ERROR_MSG,
        handlerActions.sendMessageRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            const notification = fin.desktop.Notification.getCurrent();
            const responseAction = resActionCreator({})
            notification.sendMessage(message,()=>{
                succCb(responseAction);
            });
        }
    )
}

//http://cdn.openfin.co/jsdocs/beta/fin.desktop.Notification.html#sendMessageToApplication
export async function sendMessageToApplication(action:Action<types.NotificationSendMsgToAppPayload>):Promise<Action<types.NotificationSendMsgToAppResPayload>>{

    const { payload: {
        message
    }} = action;

    return createAsyncFun<types.NotificationSendMsgToAppPayload,types.NotificationSendMsgToAppResPayload>(
        action,
        NOTIFICATION_SEND_MSG_TO_APP_ERROR_MSG,
        handlerActions.sendMessageToApplicationRes,
        (fin,action,resActionCreator,succCb,errCb)=>{
            const notification = fin.desktop.Notification.getCurrent();
            const responseAction = resActionCreator({})
            notification.sendMessageToApplication(message,()=>{
                succCb(responseAction);
            });
        }
    )
}