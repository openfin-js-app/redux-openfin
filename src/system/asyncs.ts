import {Action} from 'redux-actions';
import * as types from "./types";
import * as handlerActions from './actions/handlerActionCreator';

import wrapAsyncFun from '../utils/wrapAsyncFun'

export async function getMachineId(action:Action<types.GetMachineIdPayload>):Promise<Action<types.GetMachineIdResPayload>>{
    return wrapAsyncFun<types.GetMachineIdPayload,types.GetMachineIdResPayload>(
        action,
        handlerActions.getMachineIdRes,
        async (fin)=>{
            const id = await fin.System.getMachineId();
            return handlerActions.getMachineIdRes({id});
        }
    );
}

export async function getDeviceUserId(action:Action<types.GetDeviceUserIdPayload>):Promise<Action<types.GetDeviceUserIdResPayload>>{
    return wrapAsyncFun<types.GetDeviceUserIdPayload,types.GetDeviceUserIdResPayload>(
        action,
        handlerActions.getDeviceUserIdRes,
        async (fin)=>{
            const id = await fin.System.getDeviceUserId();
            return handlerActions.getDeviceUserIdRes({id});
        }
    );
}

export async function getMonitorInfo(action:Action<types.GetMonitorInfoPayload>):Promise<Action<types.GetMonitorInfoResPayload>>{
    return wrapAsyncFun<types.GetMonitorInfoPayload,types.GetMonitorInfoResPayload>(
        action,
        handlerActions.getMonitorInfoRes,
        async (fin)=>{
            const monitorInfo = await fin.System.getMonitorInfo();
            return handlerActions.getMonitorInfoRes(monitorInfo);
        }
    );
}

export async function getVersion(action:Action<types.GetVersionPayload>):Promise<Action<types.GetVersionResPayload>>{
    return wrapAsyncFun<types.GetVersionPayload,types.GetVersionResPayload>(
        action,
        handlerActions.getVersionRes,
        async (fin)=>{
            const version = await fin.System.getVersion();
            return handlerActions.getVersionRes({version});
        }
    );
}

export async function getHostSpecs(action:Action<types.GetHostSpecsPayload>):Promise<Action<types.GetHostSpecsResPayload>>{
    return wrapAsyncFun<types.GetHostSpecsPayload,types.GetHostSpecsResPayload>(
        action,
        handlerActions.getHostSpecsRes,
        async (fin)=>{
            const info = await fin.System.getHostSpecs();
            return handlerActions.getHostSpecsRes(info);
        }
    );
}


export async function getEnvironmentVariable(action:Action<types.GetEnvironmentVariablePayload>)
    :Promise<Action<types.GetEnvironementVariableResPaylod>>
{
    const { env } = action.payload;
    return wrapAsyncFun<types.GetEnvironmentVariablePayload,types.GetEnvironementVariableResPaylod>(
        action,
        handlerActions.getEnvironmentVariableRes,
        async (fin)=>{
            const value = await fin.System.getEnvironmentVariable(env);
            return handlerActions.getEnvironmentVariableRes({env,value});
        }
    );
}

export async function clearCache(action:Action<types.ClearCachePayload>):Promise<Action<types.ClearCacheResPayload>>{
    const  options  = action.payload;
    return wrapAsyncFun<types.ClearCachePayload,types.ClearCacheResPayload>(
        action,
        handlerActions.clearCacheRes,
        async (fin)=>{
            await fin.System.clearCache(options);
            return handlerActions.clearCacheRes({});
        }
    );
}