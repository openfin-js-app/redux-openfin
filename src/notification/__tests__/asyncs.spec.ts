jest.mock('../../utils/createAsyncFun');
const createAsyncFun = require('../../utils/createAsyncFun');
import * as actions from '../actions';
import * as asyncs  from '../asyncs';

jest.useFakeTimers();

describe('Notification asyncs',()=>{

    afterEach(()=>{
        jest.resetAllMocks();
    });

    // it('createNotification',async ()=>{
    //     const fin = {
    //         desktop:{
    //             Notification : jest.fn((options,succCb,errCb)=>{
    //                 setTimeout(()=>{
    //                     succCb();
    //                 },0);
    //                 return {};
    //             })
    //         }
    //     };
    //     const succCb = jest.fn();
    //     const errCb = jest.fn();
    //
    //     createAsyncFun.default=jest.fn(
    //         (action,ERROR_MSG,resActionCreator,finCb)=>{
    //             finCb(fin,action,resActionCreator,succCb,errCb);
    //         }
    //     );
    //     await asyncs.createNotification(
    //         actions.createNotification({
    //             ignoreMouseOver:false,
    //             url:'url',
    //             message:'message',
    //             onClick:()=>{},
    //             onClose:()=>{},
    //             onDismiss:()=>{},
    //             onError:()=>{},
    //             onMessage:()=>{},
    //             onShow:()=>{},
    //             timeout:3000,
    //             opacity:1,
    //         })
    //     );
    //     jest.runAllTimers();
    //     expect(fin.Notification).toBeCalled();
    //     expect(succCb).toHaveBeenCalled();
    //     expect(createAsyncFun).toMatchSnapshot();
    // });

});