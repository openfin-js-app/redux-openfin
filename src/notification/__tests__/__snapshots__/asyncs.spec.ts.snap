// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification asyncs createNotification 1`] = `
Object {
  "__esModule": true,
  "default": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "error": undefined,
          "payload": Object {
            "ignoreMouseOver": false,
            "message": "message",
            "onClick": [Function],
            "onClose": [Function],
            "onDismiss": [Function],
            "onError": [Function],
            "onMessage": [Function],
            "onShow": [Function],
            "opacity": 1,
            "timeout": 3000,
            "url": "url",
          },
          "type": "REDUX_OPENFIN_REQ::CREATE_NOTIFICATION",
        },
        "OpenFin API call Notification.constructor failed.",
        [Function],
        [Function],
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
}
`;
